### ISSUE TITLE
Weird and wonderful

### ISSUE THREAD

**Issue by computerscot** [net4people/bbs#279]
We've seen a tightening of restrictions this year in Iran, Russia, and other countries. The counter-censorship measures that tend to get blocked are the ones everyone uses. If you're a firewall operator, and you see 1,000,000 people all using a certain method, then that's where you'll concentrate your efforts. 

From the individual's point of view, it makes sense to keep in reserve a few "weird and wonderful" methods -- oddball techniques that very few people use. You're more likely to survive if you're not the low-hanging fruit.

I've compiled a couple of lists of these "last resort" methods. These methods are intended to be obscure rather than speedy. One list is at the bottom of the "summary of methods" post on the old blog [here](https://cscot.pages.dev/2023/06/01/summary-of-methods):

* Iodine DNS tunnel
* pingtunnel ICMP tunnel
* V2Ray + meek
* SS over SSH over VLESS
* Shadowsocks over Restls

A second collection of "last resort" methods is on the new blog [here](https://computerscot.github.io):

* OpenVPN over Shadowsocks
* WireGuard over Xray
* Tor obfs4 bridge with iat-mode=2
* OpenVPN + Cloak
* OpenVPN + Obfs4
* WireGuard + TCP
* Obfuscated SSH

Hopefully one of these will work if you ever find all your preferred methods have been blocked.

**Comment by nametoolong** [comment]
To add another trick to the list: you can revive a vanilla Tor bridge with a fake, whitelisted SNI. Just change a single line in C Tor and you're good to go, at least in Iran. You'll also need [traffic shaping and more](https://github.com/nametoolong/oregano) in China, but that project is long dead due to changes in GnuTLS and OpenSSL.

**Comment by gusgustavo** [comment]
>Tor obfs4 bridge with iat-mode=2
I don't think this is accurate. What really works in IR is Tor Snowflake - https://metrics.torproject.org/userstats-bridge-combined.html?start=2023-05-30&end=2023-08-28&country=ir

**Comment by wkrp** [comment]
> you can revive a vanilla Tor bridge with a fake, whitelisted SNI. Just change a single line in C Tor and you're good to go, at least in Iran.

More details on this point:

https://ntc.party/t/tor-relay-tls-clienthello-sni-fingerprinting/3922

> Iranian [ParsOnline](https://parsonline.net/) ISP has a regexp filter for the domain generation algorithm used in Tor for ClientHello SNI field.
> `www\.[a-z]{4,25}\.com`
>
> https://gitlab.torproject.org/tpo/core/tor/-/blob/21b3397f9b0803134bc458b83cd161de259887fd/src/lib/tls/tortls_openssl.c#L1082
>
> You can connect to the majority of Tor relays, but the connection would freeze if this type of SNI is used along with ClientHello size <= 256 bytes.
> Changing the domain name even slightly, like .com to .org allows to connect to Tor Relay.
This filter seem to be applied only to known Relays IP addresses in Tor network.

**Comment by wkrp** [comment]
> * Iodine DNS tunnel

For use cases involving Iodine, consider [dnstt](https://github.com/net4people/bbs/issues/30): it likely has better performance, and can work with DoH and DoT resolvers.

**Comment by testcaoy7** [comment]
DNS and ICMP tunnels are very slow, but I agree that they can be used in case of dire need.
Other tools worth people's attention:

Stunnel - Universal TLS wrapper with socks5 built-in

Tinc - A peer-to-peer mesh-like VPN with long history, the first release was made in late 90s.

SoftEther - A TLS VPN but more like a ethernet wire emulator. You can even run non-IP protocol on it. It comes with a standalone GUI server manager. (Important: fresh installed SoftEther contains certificates that immediately trigger blocking. Either configure it to use your own certificate under SSH or connect GUI server manager via an existing proxy to configure its certificate.)

**Comment by alidxdydz** [comment]
@nametoolong 

Since you mentioned shape shifting, I wonder if anyone has ever used "Dust" [v1](https://github.com/blanu/Dust/tree/master/historical/v1) and [v2](https://github.com/blanu/Dust/blob/master/historical/haskell2014/Dust/README) and its ongoing successor project ["ShapeShifter" Transports](https://github.com/OperatorFoundation/shapeshifter-transports).
Dust's [presentation](https://youtu.be/u0iueUlY6dQ) looked promising when it was introduced and it makes me wonder why we do not hear about this project very often!

**Comment by alidxdydz** [comment]
@nametoolong 

Since you mentioned traffic shaping, I wonder if anyone has ever used "Dust" [v1](https://github.com/blanu/Dust/blob/master/historical/v1/README) and [v2](https://github.com/blanu/Dust/blob/master/historical/haskell2014/Dust/README) and its ongoing successor project ["ShapeShifter" Transports](https://github.com/OperatorFoundation/shapeshifter-transports).
Dust's [presentation](https://youtu.be/u0iueUlY6dQ) looked promising when it was introduced and it makes me wonder why we do not hear about this project very often!

**Comment by nametoolong** [comment]
I did some experiments with Dust. Its concept was really fancy; it was complicated and not as performant as ScrambleSuit; [its shaping layer was a parrot that researchers have doubt on](https://github.com/blanu/Dust/issues/22#issuecomment-603648691). The community later moved on and made heavier use of tunneling-based solutions.

I believe protocol mimicry has its advantage. With mimicry, a user can easily find an unblocked spot in the feature space. For large-scale deployment, however, everyone wants stability. Maybe you remember constantly changing protocols a decade ago? Mimicry encourages constant cat-and-mouse game as despised by proxy operators, who generally prefer the higher adversarial computational cost associated with identifying an encrypted tunnel. Since the cat-and-mouse game never stops, protocol mimicry might have a second spring in the future when circumvention becomes more decentralized.

**Comment by mobilelifeful** [comment]
Kamino closed and cloned this issue to [mobilelifeful/bbs](https://github.com/mobilelifeful/bbs/issues/3)

**Comment by mobilelifeful** [comment]
Kamino closed and cloned this issue to [mobilelifeful/bbs](https://github.com/mobilelifeful/bbs/issues/4)

**Comment by computerscot** [comment]
> For use cases involving Iodine, consider [dnstt](https://github.com/net4people/bbs/issues/30): it likely has better performance, and can work with DoH and DoT resolvers.

Is DNS-over-HTTPS blocked in China?

**Comment by wkrp** [comment]
> Is DNS-over-HTTPS blocked in China?

In general, yes, DoH is blocked in China (see e.g. [Section V-E of this](https://www.ndss-symposium.org/ndss-paper/auto-draft-123/)), but dnstt can also run in a plain UDP/53 mode like Iodine. It's not good for circumvention except against a naive censor, because the hostname of the dnstt server is exposed in DNS messages, but the same concern exists with Iodine on UDP/53. The benefits of dnstt is that it can go a lot [faster](https://www.bamsoftware.com/software/dnstt/performance.html#download-20200430) and it is [always encrypted](https://www.bamsoftware.com/software/dnstt/protocol.html#crypto) to prevent inspection or tampering by the intermediate resolver.

**Comment by computerscot** [comment]
@wkrp Thanks. I have added `dnstt` to the "weird and wonderful" list as a modern replacement for Iodine.
